src/Reflection/InitializerExprTypeResolver.php

```diff
@@ -506,6 +506,10 @@ public function resolveConcatType(Type $left, Type $right): Type
 
 		$leftNumericStringNonEmpty = TypeCombinator::remove($leftStringType, new ConstantStringType(''));
 		if ($leftNumericStringNonEmpty->isNumericString()->yes()) {
+			$allowedRightPattern = $left->isInteger()->yes()
+				? '#^(\d+|\d+.\d+)([eE][+-]?\d+)?$#' // non-negative integer, float or scientific string
```

Comment:
Maybe this can even be "doesn't start with minus sign" and is_numeric() so we don't re-implement the logic which defines what a numeric string is?